Dart/src/com/jetbrains/lang/dart/ide/formatter/DartPreFormatProcessor.java

```diff
@@ -0,0 +1,133 @@
+import com.intellij.psi.FileViewProvider;
+import com.intellij.psi.PsiDocumentManager;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.codeStyle.CodeStyleSettings;
+import com.intellij.psi.codeStyle.CodeStyleSettingsManager;
+import com.intellij.psi.impl.source.codeStyle.PreFormatProcessor;
+import com.jetbrains.lang.dart.DartLanguage;
+import com.jetbrains.lang.dart.analyzer.DartAnalysisServerService;
+import com.jetbrains.lang.dart.sdk.DartSdk;
+import com.jetbrains.lang.dart.sdk.DartSdkGlobalLibUtil;
+import org.dartlang.analysis.server.protocol.SourceEdit;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.*;
+import java.util.Collection;
+import java.util.List;
+
+public class DartPreFormatProcessor implements PreFormatProcessor {
+  public static Key<Boolean> FORMAT_MARK = new Key<Boolean>("FORMAT_MARK");
+  public static Boolean FORMAT_MARKER = Boolean.TRUE;
+
+  private CodeStyleSettings mySettings;
+  private PsiFile myFile;
+  private Project myProject;
+  private int selectionOffset;
+
+  @NotNull
+  @Override
+  public TextRange process(@NotNull ASTNode element, @NotNull TextRange range) {
+    // TODO return range if not in a paste operation
+    final PsiElement psiElement = element.getPsi();
+    if (psiElement == null) return range;
+
+    if (!psiElement.getLanguage().is(DartLanguage.INSTANCE)) return range;
+
+    if (!psiElement.isValid()) return range;
+    myFile = psiElement.getContainingFile();
+    if (myFile == null || !myFile.isWritable()) return range;
+
+    myProject = psiElement.getProject();
+    mySettings = CodeStyleSettingsManager.getSettings(myProject);
+    final FileViewProvider viewProvider = myFile.getViewProvider();
+    final Document document = viewProvider.getDocument();
+    if (document == null) return range;
+
+    final String code = reformat(document.getText(), range);
+    if (code == null) return range;
+    final RangeMarker marker = document.createRangeMarker(range.getStartOffset(), range.getEndOffset(), true);
+    document.replaceString(0, document.getTextLength(), code);
+
+    element.putUserData(FORMAT_MARK, FORMAT_MARKER);
+    PsiDocumentManager documentManager = PsiDocumentManager.getInstance(myProject);
+    // Without this commit, an error is thrown by some tests.
+    // With it, some files get garbled.
+    documentManager.commitDocument(document);
+    return TextRange.create(marker.getStartOffset(), marker.getEndOffset());
+  }
+
+  public String reformat(String code, TextRange range) {
+    final DartAnalysisServerService.FormatResult formatResult;
+    try {
+      // While running a paste operation the contents of the VFS are not in sync with the document.
+      // Put the document content in a temporary file to be reformatted.
+      String parentPath = parentFilePath();
+      if (parentPath == null) return null;
+      final File tempFile = FileUtil.createTempFile(new File(parentPath), "fmt", ".dart");
```

Comment:
Disk operations are potentially slow and look like an overhead for this task. DAS can work with content that is not saved on disk. DartAnalysisServerService#updateFilesContent sends current contents of all unsaved Documents (modified since previous sync) to the server. Other options is to extend server API so that its edit_format() could also accept file contents instead of a file path.